### DISCUSSION TITLE
Q&A Lab1

### DISCUSSION THREAD

**Discussion by rolandkasa** [rolandkasa/TastyWeb#2] [General]
Hi all,

Here you can post your questions if you have any regarding the "Home Work" or any technical difficulties that you face.

**Comment by rolandkasa** [comment]
Un lucru ce am uitat sa va zic.

O sa aveti probleme de CORS.  Asta inseamna ca API-ul nu-l recunoaste interfata WEB ca un consumator acceptat.
Prietenul nostru mozzilla developers: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS

Va trebui sa faceti niste modificari in back-end -> Startup.cs

using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.HttpsPolicy;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Microsoft.OpenApi.Models;
using TastyAPI.Domain.Mapping;
using TastyAPI.Repository;
using WebSocketOptions = Microsoft.AspNetCore.Http.Connections.WebSocketOptions;

namespace TastyAPI
{
    public class Startup
    {
        readonly string MyAllowSpecificOrigins = "_myAllowSpecificOrigins";

        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            Console.WriteLine(Configuration.GetConnectionString("DefaultConnection"));
            services.AddDbContext<TastyDbContext>(options =>
                options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")));

            services.AddScoped<IProductRepository, ProductRepository>();
            services.AddScoped<IOrderRepository, OrderRepository>();
            services.AddScoped<ICategoryRepository, CategoryRepository>();

            services.AddAutoMapper(typeof(TastyMapping));
            services.AddCors(options =>
                    {
                        options.AddPolicy(name: MyAllowSpecificOrigins,
                                        builder =>
                                        {
                                            builder.WithOrigins("http://localhost:3000")
                                                .AllowAnyHeader()
                                                .AllowAnyMethod();
                                        });
                    });
            services.AddSwaggerGen(options =>
            {
                options.SwaggerDoc("v1",
                    new OpenApiInfo
                    {
                        Title = "Tasty API",
                        Version = "1",
                        Description = "Demo Food Order API"
                    });
                //include xml comments
                var xmlCommentFile = $"{Assembly.GetEntryAssembly().GetName().Name}.xml";
                var xmlCommentsFullPath = Path.Combine(AppContext.BaseDirectory, xmlCommentFile);
                options.IncludeXmlComments(xmlCommentsFullPath);
            });

            services.AddControllers();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseHttpsRedirection();
            app.UseSwagger();
            app.UseSwaggerUI(options =>
            {
                options.SwaggerEndpoint("swagger/v1/swagger.json", "Tasty API");
                options.RoutePrefix = string.Empty;
            });
            app.UseRouting();
            app.UseCors(MyAllowSpecificOrigins);
            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
            });

            using (var serviceScope = app.ApplicationServices.GetService<IServiceScopeFactory>().CreateScope())
            {
                var context = serviceScope.ServiceProvider.GetRequiredService<TastyDbContext>();
                context.Database.Migrate();
            }
        }
    }
}

**Comment by cristianpuiu** [comment]
Salut, 

Am o problema in ceea ce priveste preluarea datelor de la API cred. Am schimbat ceea ce ai psu tu mai sus si adresa hostului in call-urile catre API si nimic. Atasez doua screenshot-uri, unul in care am pus call-urile catre API si unul de ansamblu.

![overview](https://user-images.githubusercontent.com/83396863/116533737-68438080-a8ea-11eb-8ae1-7f94ad103d1b.jpg)
![API-calls](https://user-images.githubusercontent.com/83396863/116533741-68dc1700-a8ea-11eb-82cf-0b1cfd752e62.jpg)

**Reply by rolandkasa** [reply]
@cristianpuiu  Vad ca ai folosit portul 44309 care foloseste protocolul https. Ori il schimbi in FE pe https, or schimbi pe portul 7520.

**Reply by rolandkasa** [reply]
@cristianpuiu Ca si un insight, eu folosesc portul 8080, pentru ca l-am dockerizat API-ul. Daca te intereseaza partea de containerizare as putea sa te ajut cu niste informatii, si fisierul Docker.

**Reply by cristianpuiu** [reply]
@rolandkasa multumesc, am remediat. Poți sa îmi trimiți te rog info de care ai menționat.

**Reply by rolandkasa** [reply]
Salut @cristianpuiu!

Partea de instalare: https://docs.docker.com/docker-for-windows/install/
Partea de intro: https://docs.docker.com/get-started/
Partea de docker-compose: https://docs.docker.com/compose/gettingstarted/

Fisierul de docker: 
FROM mcr.microsoft.com/dotnet/core/sdk:3.1 AS build-env
WORKDIR /app


COPY . ./
RUN dotnet restore --runtime=ubuntu.14.04-x64


COPY . ./
RUN dotnet build -o out
RUN dotnet publish -c Release -o out


FROM mcr.microsoft.com/dotnet/core/aspnet:3.1
WORKDIR /app
COPY --from=build-env /app/out .

ENTRYPOINT ["dotnet", "TastyAPI.dll"]


Fisierul de docker-compose:
version: "3.3"
services:
  db:
    image: "mcr.microsoft.com/mssql/server:2019-latest"
    container_name: mssql1
    user: root
    environment:
      SA_PASSWORD: "test!123"
      ACCEPT_EULA: "Y"
    ports:
      - "1433:1433"
    networks: 
      backend:
        aliases:
          - mssql1
  web:
    build: .
    ports:
      - "8080:80"
    networks: 
      - backend
    depends_on:
      - db
    restart: always
networks:
  backend:
    driver: bridge

Schimbari in appsettings.json:
{
  "ConnectionStrings": {
    "DefaultConnection": "Server=mssql1;Initial Catalog=Tasty;User=sa;Password=test!123;"
  } 
}

Happy coding!

**Comment by SimonaRadulea** [comment]
Salut, nu mai știu exact cum ar trebui sa trimitem tema, daca ai putea sa ne spui din nou. :D

**Reply by rolandkasa** [reply]
Salut @SimonaRadulea,

Presupun ca ati facut Fork din repo-ul acesta.
Dupa ce push-uiti modificatiile facute de voi pe un branch oarecare in fork, aveti posibilitatea sa creati un Pull Request inapoi catre repo-ul de baza. Acolo selectati developul ca si Base Branch.

Daca nu ati folosit functionalitatea de Fork, si ati creat un branch in repo-ul acesta, atunci pur si simplu creati un Pull Requast cu develop ca si base branch.

Ca si referinta adaug si un articol.
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork

Happy coding!

**Reply by SimonaRadulea** [reply]
Super, mersi!
